File: src/javascript/lib/map.js
import SpecialForms from './kernel/special_forms';
import * as Core from './core';

function __new__(){
  return SpecialForms.map({});
}

function keys(map){
  return Core.Functions.get_object_keys(map);
}

function size(map){
  return keys(map).length;
}

function to_list(map){
  let map_keys = keys(map);
  let list = [];

  for(let key of map_keys){
    list.push(new Core.Tuple(key, map[key]));
  }

  return Core.List(...list);
}

function values(map){
  let map_keys = keys(map);
  let list = [];

  for(let key of map_keys){
    list.push(map[key]);
  }

  return Core.List(...list);
}

function from_struct(struct){
  let map = Object.assign({}, struct);
  delete map[Symbol.for("__struct__")];

  return SpecialForms.map(map);
}

function __delete__(map, key){
  let new_map = Object.assign({}, map);

  delete new_map[key];

  return SpecialForms.map(new_map);
}

function drop(map, keys){
  let new_map = Object.assign({}, map);

  for(let key of keys){
    delete new_map[key];
  }

  return SpecialForms.map(new_map);
}


function equal__qmark__(map1, map2){
  return map1 === map2;
}

function fetch__emark__(map, key){
  if(key in map){
    return map[key]
  }

  throw new Error("Key not found.");
}

function fetch(map, key){
  if(key in map){
    return new Core.Tuple(Symbol.for("ok"), map[key]);
  }

  return Symbol.for("error");
}

function has_key__qmark__(map, key){
  return key in map;
}

function merge(map1, map2){
  return SpecialForms.map_update(map1, map2);
}

function split(map, keys){
  let split1 = {};
  let split2 = {};

  for(let key of Core.Functions.get_object_keys(map) ){
    if(keys.indexOf(key) > -1){
      split1[key] = map[key];
    }else{
      split2[key] = map[key];
    }
  }

  return new Core.Tuple(
    SpecialForms.map(split1),
    SpecialForms.map(split2)
  );
}

function take(map, keys){
  let split1 = {};

  for(let key of Core.Functions.get_object_keys(map) ){
    if(keys.indexOf(key) > -1){
      split1[key] = map[key];
    }
  }

  return SpecialForms.map(split1);
}

function drop(map, keys){
  let split1 = {};

  for(let key of Core.Functions.get_object_keys(map) ){
    if(keys.indexOf(key) === -1){
      split1[key] = map[key];
    }
  }

  return SpecialForms.map(split1);
}

function put_new(map, key, value){
  if(key in map){
    return map;
  }

  let new_map = Object.assign({}, map);
  new_map[key] = value;

  return SpecialForms.map(new_map);
}

function put_new_lazy(map, key, fun){
  if(key in map){
    return map;
  }

  let new_map = Object.assign({}, map);
  new_map[key] = fun();

  return SpecialForms.map(new_map);
}

function get_and_update(map, key, fun){
  if(key in map){
    return map;
  }

  let new_map = Object.assign({}, map);
  new_map[key] = fun(map[key]);

  return SpecialForms.map(new_map);
}

function pop_lazy(map, key, fun){
  if(!key in map){
    return new Core.Tuple(fun(), map);
  }

  let new_map = Object.assign({}, map);
  let value = fun(new_map[key]);
  delete new_map[key];

  return new Core.Tuple(value, new_map);
}


function pop(map, key, _default = null){
  if(!key in map){
    return new Core.Tuple(_default, map);
  }

  let new_map = Object.assign({}, map);
  let value = new_map[key];
  delete new_map[key];

  return new Core.Tuple(value, new_map);
}

function get_lazy(map, key, fun){
  if(!key in map){
    return fun();
  }

  return fun(map[key]);
}


function get(map, key, _default = null){
  if(!key in map){
    return _default;
  }

  return map[key];
}

function put(map, key, val){
  let new_map = Object({}, map);
  new_map[key] = val;

  return SpecialForms.map(new_map);
}

function update__emark__(map, key, fun){
  if(!key in map){
    throw new Error("Key not found");
  }

  let new_map = Object({}, map);
  new_map[key] = fun(map[key]);

  return SpecialForms.map(new_map);
}

function update(map, key, initial, fun){
  let new_map = Object({}, map);

  if(!key in map){
    new_map[key] = initial;
  }else{
    new_map[key] = fun(map[key]);
  }

  return SpecialForms.map(new_map);
}


export default {
  new: __new__,
  keys,
  size,
  to_list,
  values,
  from_struct,
  delete: __delete__,
  drop,
  equal__qmark__,
  fetch__emark__,
  fetch,
  has_key__qmark__,
  split,
  take,
  put_new,
  put_new_lazy,
  get_and_update,
  pop_lazy,
  pop,
  get_lazy,
  get,
  put,
  update__emark__,
  update
}

File: priv/std_lib/map.ex
defmodule ElixirScript.Map do
  @moduledoc false

  def new() do
    %{}
  end

  def keys(map) do
    ElixirScript.Bootstrap.Functions.get_object_keys(map)
  end

  def size(map) do
    keys(map).length
  end

  def to_list(map) do
    do_to_list(map, [])
  end

  def do_to_list(map, list) do
    case size(map) do
      0 ->
        list
      _ ->
        key = hd(keys(map))
        value = map[key]
        do_to_list(Map.delete(map, key), list ++ [{key, value}])
    end
  end

  def values(map) do
    JS.Object.values(map)
  end

  def from_struct(struct) do
    struct
    |> ElixirScript.Bootstrap.Functions.class_to_obj
    |> delete(:__struct__)
  end

  def delete(map, key) do
    map
    |> ElixirScript.Bootstrap.Functions.delete_property_from_map(key)
  end

  def equal?(map1, map2) do
    map1 === map2
  end

  def fetch!(map, key) do
    case key in keys(map) do
      true ->
        map[key]
      false ->
        raise "#{key} not found in map"
    end
  end

  def fetch(map, key) do
    case key in keys(map) do
      true ->
        { :ok, map[key] }
      false ->
        :error
    end
  end

  def has_key?(map, key) do
    key in keys(map)
  end

  def merge(map1, map2) do
    JS.Object.assign(%{}, map1, map2)
    |> JS.Object.freeze()
  end

  def split(map, keys) do
    do_split(map, keys, { %{}, %{} })
  end

  defp do_split(_, [], split_tuple) do
    split_tuple
  end

  defp do_split(map, keys, { key_map, non_key_map }) do
    key = hd(keys)

    new_split_tuple = case key in keys(map) do
      true ->
        { Map.put(key_map, key, map[key]), non_key_map }
      false ->
        { key_map, Map.put(non_key_map, key, map[key]) }
    end

    do_split(map, tl(keys), new_split_tuple)
  end

  def take(map, keys) do
    {key_map, _} = split(map, keys)
    key_map
  end

  def drop(map, keys) do
    {_, non_key_map} = split(map, keys)
    non_key_map
  end

  def put_new(map, key, value) do
    case key in keys(map) do
      true ->
        map
      false ->
        Map.put(map, key, value)
    end
  end

  def put_new_lazy(map, key, func) do
    case key in keys(map) do
      true ->
        map
      false ->
        Map.put(map, key, func.())
    end
  end

  def put(map, key, value) do
    ElixirScript.Bootstrap.Functions.add_property_to_map(map, key, value)
  end

  def get(map, key) do
    get(map, key, nil)
  end

  def get(map, key, default_value) do
    case key in keys(map) do
      true ->
        map[key]
      false ->
        default_value
    end
  end

  def get_lazy(map, key, func) do
    case key in keys(map) do
      true ->
        func.(map[key])
      false ->
        func.()
    end
  end

  def get_and_update(map, key, func) do
    case key in keys(map) do
      true ->
        { nil, map }
      false ->
        new_value = func.(map[key])
        { new_value, Map.put(map, key, new_value) }
    end
  end

  def pop(map, key) do
    pop(map, key, nil)
  end

  def pop(map, key, default_value) do
    case key in keys(map) do
      true ->
        { map[key], Map.delete(map, key) }
      false ->
        { default_value, map }
    end
  end

  def pop_lazy(map, key, func) do
    case key in keys(map) do
      true ->
        { func.(map[key]), Map.delete(map, key) }
      false ->
        { func.(), map }
    end
  end


  def update!(map, key, func) do
    case key in keys(map) do
      true ->
        Map.put(map, key, func.(map[key]))
      false ->
        raise "#{key} not found in map"
    end
  end


  def update(map, key, initial, func) do
    case key in keys(map) do
      true ->
        Map.put(map, key, func.(map[key]))
      false ->
        Map.put(map, key, initial)
    end
  end

end
